[REVIEW] Single GPU KMeans

akkamesh opened the pull request:
This contains the single GPU implementations of kmeans built using ml-prims.

Tagging @cjnolet, @teju85 and @jirikraus for review.

GPUtester commented:
Can one of the admins verify this patch?

cjnolet commented:
add to whitelist

dantegd commented:
rerun tests

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+    
+    DataT priorClusteringCost = 0;    
+    for(n_iter = 0; n_iter < max_iter; ++n_iter){
+	CUDA_CHECK(cudaStreamSynchronize(stream) );
```
Is this `cudaStreamSynchronize` necessary?

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+					      _metric,
+					      stream);
+      
+	CUDA_CHECK( cudaStreamSynchronize(stream) );
```
Is this `cudaStreamSynchronize` necessary?

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+				    cub::KeyValuePair<IndexT, DataT>*> itr(minClusterAndDistance.data(),
+									   conversion_op);
+      
+	_workspace.reserve(n_samples, stream);
```
`resize` instead of `reserve`?

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+	// Note - when sampleCountInCluster[i] is 0, newCentroid[i] is reset to 0
+      
+	// transforms int values in sampleCountInCluster to its inverse and more importantly to DataT because matrixVectorOp supports only when matrix and vector are of same type
+	_workspace.reserve(sampleCountInCluster.numElements() * sizeof(DataT), stream);
```
`resize` instead of ` reserve`?

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+
+      
+	// wait for stream to finish before sending data to other ranks...
+	CUDA_CHECK( cudaStreamSynchronize(stream) );
```
Is this `cudaStreamSynchronize` really necessary?

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+		       sqrdNorm.data(),
+		       sqrdNorm.numElements(),
+		       stream);
+	CUDA_CHECK( cudaStreamSynchronize(stream) );
```
Is this `cudaStreamSynchronize` really necessary?

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+    inertia = clusteringCost.value;
+  
+    _labelsRawData.reserve(n_samples, stream);
+    CUDA_CHECK( cudaStreamSynchronize(stream) );
```
Is this `cudaStreamSynchronize` necessary?

jirikraus commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1392 @@
+    CUDA_CHECK( cudaStreamSynchronize(stream) );
+
+    auto labels = std::move(Tensor<IndexT, 1>(_labelsRawData.data(), {n_samples}));
+    thrust::transform(thrust::cuda::par.on(stream),
```
Please use 
```cpp
ML::thrustAllocatorAdapter alloc( h.getDeviceAllocator(), stream );
auto execution_policy = thrust::cuda::par(alloc).on(stream);
```
for consistency reasons.

jirikraus commented on cuML/src/kmeans/kmeans.cuh:
```diff
@@ -0,0 +1,216 @@
+     */
+    KMeans(const ML::cumlHandle_impl &cumlHandle,
+	   int n_clusters = 8,
+	   int metric = static_cast<MLCommon::Distance::DistanceType>(0),
```
As the type definition of `MLCommon::Distance::DistanceType` is available here why can't we use `MLCommon::Distance::DistanceType` directly instead of `int`?

akkamesh replied:
We can use enum directly here and I will make the necessary change. But JFYI, at kmeans.hpp we can't use it because cython currently is passing `int` and once we expose `DistanceType` to cython then we can pass `DistanceType` uniformly across all layer.

jirikraus commented:
> This contains the single GPU implementations of kmeans built using ml-prims.
> 
> Tagging @cjnolet, @teju85 and @jirikraus for review.

@akkamesh I gave this a pass. Please tag me when it make sense to do another iteration.

dantegd commented:
rerun tests

dantegd commented:
rerun tests

akkamesh commented:
> > This contains the single GPU implementations of kmeans built using ml-prims.
> > Tagging @cjnolet, @teju85 and @jirikraus for review.
> 
> @akkamesh I gave this a pass. Please tag me when it make sense to do another iteration.

Thanks for the feedback @jirikraus. I have addressed all your comments -
- Most of the `cudaStreamSynchronize` was indeed unnecessary and I have removed them accordingly.
- Fixed the use of `ML::detail::streamSyncer`.
- Exposed `MLCommon::Distance::DistanceType` directly in kmeans.cuh.
- Update thrust calls to follow [https://github.com/rapidsai/cuml/blob/branch-0.7/cuML/DEVELOPER_GUIDE.md#using-thrust](https://github.com/rapidsai/cuml/blob/branch-0.7/cuML/DEVELOPER_GUIDE.md#using-thrust)

akkamesh commented on python/cuml/cluster/kmeans.pyx:
```diff
@@ -268,53 +313,71 @@ class KMeans:
+            fit_predict(
+                handle_[0],
+                <int> self.n_clusters,         # n_clusters                
+	        <int> 0,                       # distance metric as L2: @todo - support other metrics
```
I'm with you on this @JohnZed. We can make that change once DistanceType from ML prims is exposed in cython.

akkamesh commented:
> I'm not the right person to judge the cuda stream changes, but the results for kmeans++ init seem significantly improved. Clustering with kmeans++ now ends up with almost the same results as random almost all the time (a little surprising actually). If you do an initializer-only torture test like the attached one, sklearn still does a bit better and you see some cuml kmeans++ failures on some fairly clean cases. But it's much more consistent than the previous rev for sure. Maybe we can work on a follow-up change with some unit-level/functional-level tests for kmeans++ initialization specifically to repro these failure cases?
> 
> [compare-kmeans-loop-updated-show-deviations.pdf](https://github.com/rapidsai/cuml/files/3144749/compare-kmeans-loop-updated-show-deviations.pdf)

Thanks @JohnZed . I wouldn't tag it as a failure just yet because cuml is using scalable-kmeans++ (or kmeans||) for initialization and scikit-learn is using simple k-means++. The comparison here is between two different algorithms which IMHO is not a fair comparison. In any case, I agree that we should have strong unit-level/functional-level tests in place.

cjnolet commented:
> Intentional whitespace changes or accidental?

Unfortunately, this keeps happening because we have not established standard formattings for the c++ code yet. 

I thought I had created an issue for this but I'm having trouble finding it now. A good step to help with this might be to export an eclipse formatter xml that can be checked into the codebase so that everyone is using the same formatting settings.

cjnolet commented:
I created #560 to reflect consistent formatting across cuML contributors.

dantegd commented:
rerun tests

akkamesh commented:
> Why does this default to 1 ("euclidean" / true l2) while transform defaults to 0 (squared l2)

The difference in the metric is intentional. Its the otherway around, transform defaults to '1'. The transform to cluster-distance space in scikit-learn returns euclidean distance (which in cuml is `MLCommon::Distance::DistanceType::EucExpandedL2Sqrt`) while the distance metric used in fit/predict is `MLCommon::Distance::DistanceType::EucExpandedL2`. Does it make sense now?

The definition of `DistanceType` is [here](https://github.com/rapidsai/cuml/blob/6e4216dbcde74e71680e7178f445c3950ab28825/ml-prims/src/distance/distance.h#L29).

akkamesh commented:
@dantegd - The changes in [de0d992](https://github.com/rapidsai/cuml/pull/528/commits/de0d99279c28ded49795550de1d6379969f1aaac) now differs from the scikit-learn usage of distance metric in fit/predict and transform. 
How do we want it in cuml? Do we want to follow uniform `DistanceType` across fit/fit_predict/transform or follow scikit-learn which uses squared L2 for fit/fit_predict and L2 for transform?

I just realized that enum names in [MLCommon::Distance::DistanceType](https://github.com/rapidsai/cuml/blob/6e4216dbcde74e71680e7178f445c3950ab28825/ml-prims/src/distance/distance.h#L29) and the comments in `kmeans.pyx` are misleading.

dantegd commented:
@akkamesh lets roll back the change and match sklearn

akkamesh commented:
> @akkamesh lets roll back the change and match sklearn

Agreed! I will fix the comments in `kmeans.pyx`.

akkamesh commented on cuML/src/kmeans/kmeans-inl.cuh:
```diff
@@ -0,0 +1,1362 @@
+		   },
+		   stream);
+
+    return inRankCp;
```
Yes, it does. It just returns the tensor view without a deep copy of the raw data.

cjnolet commented:
Linking #564
